src/main/java/com/google/api/gax/grpc/OperationFuture.java

```diff
@@ -0,0 +1,372 @@
+      Future<ResponseT> dataGetterFuture,
+      CountDownLatch asyncCompletionLatch) {
+    this.startOperationFuture = startOperationFuture;
+    this.finalResultFuture = finalResultFuture;
+    this.dataGetterFuture = dataGetterFuture;
+    this.asyncCompletionLatch = asyncCompletionLatch;
+  }
+
+  private static class DataGetterRunnable<ResponseT extends Message>
+      implements Callable<ResponseT> {
+    private final ListenableFuture<Operation> startOperationFuture;
+    private final SettableFuture<ResponseT> finalResultFuture;
+    private final OperationsApi operationsApi;
+    private final Class<ResponseT> responseClass;
+    private final Duration pollingInterval;
+    private final Waiter waiter;
+    private final CountDownLatch asyncCompletionLatch;
+
+    public DataGetterRunnable(
+        ListenableFuture<Operation> startOperationFuture,
+        SettableFuture<ResponseT> finalResultFuture,
+        OperationsApi operationsApi,
+        Class<ResponseT> responseClass,
+        Duration pollingInterval,
+        Waiter waiter,
+        CountDownLatch asyncCompletionLatch) {
+      this.startOperationFuture = startOperationFuture;
+      this.finalResultFuture = finalResultFuture;
+      this.operationsApi = operationsApi;
+      this.responseClass = responseClass;
+      this.pollingInterval = pollingInterval;
+      this.waiter = waiter;
+      this.asyncCompletionLatch = asyncCompletionLatch;
+    }
+
+    @Override
+    public ResponseT call() {
+      try {
+        callImpl();
+      } finally {
+        asyncCompletionLatch.countDown();
+      }
+      return null;
+    }
+
+    public void callImpl() {
+      Operation firstOperation = null;
+      try {
+        firstOperation = startOperationFuture.get();
+        if (firstOperation.getDone()) {
+          setResultFromOperation(finalResultFuture, firstOperation, responseClass);
+          return;
+        }
+        do {
+          // TODO: switch implementation from polling to scheduled execution
+          waiter.wait(pollingInterval);
+          Operation latestOperation = operationsApi.getOperation(firstOperation.getName());
+          if (latestOperation.getDone()) {
+            if (isCancelled(latestOperation)) {
+              finalResultFuture.cancel(true);
+            } else {
+              setResultFromOperation(finalResultFuture, latestOperation, responseClass);
+            }
+            return;
+          }
+        } while (true);
+      } catch (InterruptedException e) {
+        try {
+          if (firstOperation != null) {
+            operationsApi.cancelOperation(firstOperation.getName());
+          }
+          if (!startOperationFuture.isDone()) {
+            startOperationFuture.cancel(true);
+          }
+          finalResultFuture.cancel(true);
+
+          Thread.currentThread().interrupt();
+          return;
+        } catch (Exception e2) {
+          finalResultFuture.cancel(true);
```

Comment:
I see. I agree that the interruption has a higher priority here. Just some suggestion on the code flow, how about
```
try {
  if (firstOperation != null) {
    operationsApi.cancelOperation(firstOperation.getName());
  }
  if (!startOperationFuture.isDone()) {
    startOperationFuture.cancel(true);
  }
} catch (Exception e2) {
  // Ignore
}

// Propagate interruption exception
finalResultFuture.cancel(true);
Thread.currentThread().interrupt();
```